taxcalc/tests/test_pufcsv.py

```diff
@@ -80,19 +80,21 @@ def test_agg(tests_path, puf_path):  # pylint: disable=redefined-outer-name
         raise ValueError(msg)
     # create aggregate diagnostic table using sub sample of records
     fullsample = pd.read_csv(puf_path)
-    rn_seed = 80  # to ensure two-percent sub-sample is always the same
-    subsample = fullsample.sample(frac=0.02,  # pylint: disable=no-member
+    rn_seed = 80  # to ensure sub-sample is always the same
+    subfrac = 0.02  # sub-sample fraction
+    subsample = fullsample.sample(frac=subfrac,  # pylint: disable=no-member
                                   random_state=rn_seed)
     rec_subsample = Records(data=subsample)
     calc_subsample = Calculator(policy=Policy(), records=rec_subsample)
     adt_subsample = multiyear_diagnostic_table(calc_subsample, num_years=nyrs)
     # compare combined tax liability from full and sub samples for each year
     taxes_subsample = adt_subsample.loc["Combined Liability ($b)"]
-    reltol = 0.01  # maximum allowed relative difference in tax liability
+    reltol = 0.04  # maximum allowed relative difference in tax liability
```

Comment:
Thanks @MattHJensen, that makes sense.

It does seem like we got lucky before.  For the purpose of this test, it seems like setting a random seed that generally produces small differences works well enough.  Then, every time a new data set is released we would update the tolerance and the seed.

If we chose a different random seed every time then there would be cases where an unlucky sample is drawn and the test would fail even though nothing is wrong.  This problem could be solved by running several iterations and as long as x% of the iterations pass then the test is passed.  But that may be overkill for this problem.  I re-wrote the test function to do this and it takes about a minute and a half to run versus 45 seconds for the original test function.

```
@pytest.mark.requires_pufcsv
def test_agg(tests_path, puf_path):  # pylint: disable=redefined-outer-name
    """
    Test Tax-Calculator aggregate taxes with no policy reform using
    the full-sample puf.csv and a two-percent sub-sample of puf.csv
    """
    # pylint: disable=too-many-locals,too-many-statements
    puf = pd.read_csv(puf_path)
    
    ...    
    
    # create aggregate diagnostic table using sub sample of records
    fullsample = puf.copy()
    use_weights = False
    rn_seed = 0
    msgs = []
    n_tests = 50
    for i in range(0, n_tests):
        msg = ""
        rn_seed = np.random.randint(0, 10000)  # to ensure sub-sample is always the same
        subfrac = 0.02  # sub-sample fraction
        if use_weights:  # normalize weights for sample
            fullsample['weights'] = fullsample.s006/fullsample.s006.sum()
            weights = 'weights'
        else: # set weights to None and pandas uses uniform dist to select samp
            weights = None
        subsample = fullsample.sample(frac=subfrac,  # pylint: disable=no-member
                                      random_state=rn_seed,
                                      weights=weights,
                                      axis=0)
        weight_prop = subsample.s006.sum()/fullsample.s006.sum()
        rec_subsample = Records(data=subsample)
        calc_subsample = Calculator(policy=Policy(), records=rec_subsample)
        adt_subsample = multiyear_diagnostic_table(calc_subsample, num_years=nyrs)
        # compare combined tax liability from full and sub samples for each year
        taxes_subsample = adt_subsample.loc["Combined Liability ($b)"]
        reltol = 0.04  # maximum allowed relative difference in tax liability
        if not np.allclose(taxes_subsample, taxes_fullsample,
                           atol=0.0, rtol=reltol):
            msg = 'PUFCSV AGG RESULTS DIFFER IN SUB-SAMPLE AND FULL-SAMPLE\n'
            msg += 'WHEN subfrac = {:.3f} and reltol = {:.4f} '
            msg += 'and weight prop = {:.4f} and seed = {:.0f}\n'
            msg = msg.format(subfrac, reltol, weight_prop, rn_seed)
            it_sub = np.nditer(taxes_subsample, flags=['f_index'])
            it_all = np.nditer(taxes_fullsample, flags=['f_index'])
            while not it_sub.finished:
                cyr = it_sub.index + calc_start_year
                tax_sub = float(it_sub[0])
                tax_all = float(it_all[0])
                reldiff = abs(tax_sub - tax_all) / abs(tax_all)
                if reldiff > reltol:
                    msgstr = ' year,sub,full,reldif= {}\t{:.2f}\t{:.2f}\t{:.4f}\n'
                    msg += msgstr.format(cyr, tax_sub, tax_all, reldiff)
                it_sub.iternext()
                it_all.iternext()

        if len(msg) > 0:
            msgs.append(msg)
    fail_rate = (len(msgs)/float(n_tests))*100
    if fail_rate > 0.3:
        fail_msg = "Failure rate: {:.2f}%".format(fail_rate)
        raise ValueError('\n'.join([fail_msg] + msgs))
```